File: tools/arduino-swift/swift/libs/socket/socket.swift
// ============================================================
// ArduinoSwift - Socket Library (TCP)
// File: swift/libs/socket/socket.swift
//
// Thin, Embedded Swift-friendly wrappers over the C ABI in arduino/libs/socket.
// No external Swift modules required.
//
// IMPORTANT: This file assumes U8 is defined in swift/core/Types.swift.
// ============================================================

public final class TCPClient: @unchecked Sendable {
    private var h: UnsafeMutableRawPointer?

    public init() {
        self.h = _socket_client_create()
    }

    internal init(adopt handle: UnsafeMutableRawPointer) {
        self.h = handle
    }

    deinit {
        _socket_client_destroy(h)
        h = nil
    }

    public func connect(host: String, port: UInt16) -> Bool {
        host.withCString { cstr in
            _socket_client_connect(h, cstr, port)
        }
    }

    public func available() -> Int {
        Int(_socket_client_available(h))
    }

    public func read(into ptr: UnsafeMutablePointer<UInt8>, max: Int) -> Int {
        Int(_socket_client_read(h, ptr, Int32(max)))
    }

    public func write(_ ptr: UnsafePointer<UInt8>, count: Int) -> Int {
        Int(_socket_client_write(h, ptr, Int32(count)))
    }

    public func write(_ bytes: [U8]) -> Int {
        bytes.withUnsafeBufferPointer { buf in
            guard let base = buf.baseAddress else { return 0 }
            return write(base, count: buf.count)
        }
    }

    public func writeString(_ s: String) -> Int {
        if s.isEmpty { return 0 }
        return write(Array(s.utf8))
    }

    public func connected() -> Bool {
        _socket_client_connected(h)
    }

    public func close() {
        _socket_client_stop(h)
    }
}

public final class TCPServer: @unchecked Sendable {
    private var h: UnsafeMutableRawPointer?

    public init(port: UInt16) {
        self.h = _socket_server_create(port)
    }

    deinit {
        _socket_server_destroy(h)
        h = nil
    }

    @discardableResult
    public func begin() -> Bool {
        _socket_server_begin(h)
    }

    public func accept() -> TCPClient? {
        guard let ch = _socket_server_accept(h) else { return nil }
        return TCPClient(adopt: ch)
    }
}

File: tools/arduino-swift/arduino/libs/socket/socket.cpp
// ============================================================
// ArduinoSwift - Socket Library (TCP) C ABI Bridge
// File: arduino/libs/socket/socket.cpp
// ============================================================
#include "socket.h"

#include <WiFi.h>

struct AS_SocketServer {
  WiFiServer server;
  bool begun;
  explicit AS_SocketServer(uint16_t port) : server(port), begun(false) {}
};

struct AS_SocketClient {
  WiFiClient client;
};

socket_server_t socket_server_create(uint16_t port) {
  return (socket_server_t)(new AS_SocketServer(port));
}

void socket_server_destroy(socket_server_t server) {
  if (!server) return;
  delete (AS_SocketServer*)server;
}

bool socket_server_begin(socket_server_t server) {
  if (!server) return false;
  AS_SocketServer* s = (AS_SocketServer*)server;
  if (!s->begun) {
    s->server.begin();
    s->begun = true;
  }
  return true;
}

socket_client_t socket_server_accept(socket_server_t server) {
  if (!server) return nullptr;
  AS_SocketServer* s = (AS_SocketServer*)server;
  if (!s->begun) {
    s->server.begin();
    s->begun = true;
  }

  WiFiClient c = s->server.available();
  if (!c) return nullptr;

  AS_SocketClient* out = new AS_SocketClient();
  out->client = c;
  return (socket_client_t)out;
}

socket_client_t socket_client_create(void) {
  return (socket_client_t)(new AS_SocketClient());
}

void socket_client_destroy(socket_client_t client) {
  if (!client) return;
  AS_SocketClient* c = (AS_SocketClient*)client;
  c->client.stop();
  delete c;
}

bool socket_client_connect(socket_client_t client, const char* host, uint16_t port) {
  if (!client || !host || host[0] == 0) return false;
  AS_SocketClient* c = (AS_SocketClient*)client;
  return c->client.connect(host, port);
}

int32_t socket_client_available(socket_client_t client) {
  if (!client) return 0;
  AS_SocketClient* c = (AS_SocketClient*)client;
  return (int32_t)c->client.available();
}

int32_t socket_client_read(socket_client_t client, uint8_t* out_buf, int32_t max_len) {
  if (!client || !out_buf || max_len <= 0) return 0;
  AS_SocketClient* c = (AS_SocketClient*)client;
  return (int32_t)c->client.read(out_buf, (size_t)max_len);
}

int32_t socket_client_write(socket_client_t client, const uint8_t* buf, int32_t len) {
  if (!client || !buf || len <= 0) return 0;
  AS_SocketClient* c = (AS_SocketClient*)client;
  return (int32_t)c->client.write(buf, (size_t)len);
}

bool socket_client_connected(socket_client_t client) {
  if (!client) return false;
  AS_SocketClient* c = (AS_SocketClient*)client;
  return c->client.connected();
}

void socket_client_stop(socket_client_t client) {
  if (!client) return;
  AS_SocketClient* c = (AS_SocketClient*)client;
  c->client.stop();
}